{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94a03362-3dbe-4646-8864-29c617550bb7",
   "metadata": {},
   "source": [
    "# Supplementary File 1: CLEANING DATA\n",
    "\n",
    "The purpose of the following code is to create a cleaned CSV that can be further used to predict target variables\n",
    "\n",
    "Objectives: \n",
    "(1) removing all CLASS = 1, removing conflicting information \n",
    "(2) dropping irrelevant/high empty percentage/extremely low variance variables \n",
    "(3) fixing errors in data entry in some columns (cDNA_position, CDS_position, Protein_position contain dashes and question marks)\n",
    "\n",
    "Input: clinvar_conflicting.csv, Output: clinvar_cleaned.csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0112a7ff-6533-4466-93f5-3218b26f4659",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jg/43bhw16j027fkjq2n5gb5t5m0000gn/T/ipykernel_76364/4198222604.py:6: DtypeWarning: Columns (0,38,40) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('clinvar_conflicting.csv')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 65188 entries, 0 to 65187\n",
      "Data columns (total 46 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   CHROM               65188 non-null  object \n",
      " 1   POS                 65188 non-null  int64  \n",
      " 2   REF                 65188 non-null  object \n",
      " 3   ALT                 65188 non-null  object \n",
      " 4   AF_ESP              65188 non-null  float64\n",
      " 5   AF_EXAC             65188 non-null  float64\n",
      " 6   AF_TGP              65188 non-null  float64\n",
      " 7   CLNDISDB            65188 non-null  object \n",
      " 8   CLNDISDBINCL        167 non-null    object \n",
      " 9   CLNDN               65188 non-null  object \n",
      " 10  CLNDNINCL           167 non-null    object \n",
      " 11  CLNHGVS             65188 non-null  object \n",
      " 12  CLNSIGINCL          167 non-null    object \n",
      " 13  CLNVC               65188 non-null  object \n",
      " 14  CLNVI               27659 non-null  object \n",
      " 15  MC                  64342 non-null  object \n",
      " 16  ORIGIN              65188 non-null  int64  \n",
      " 17  SSR                 130 non-null    float64\n",
      " 18  CLASS               65188 non-null  int64  \n",
      " 19  Allele              65188 non-null  object \n",
      " 20  Consequence         65188 non-null  object \n",
      " 21  IMPACT              65188 non-null  object \n",
      " 22  SYMBOL              65172 non-null  object \n",
      " 23  Feature_type        65174 non-null  object \n",
      " 24  Feature             65174 non-null  object \n",
      " 25  BIOTYPE             65172 non-null  object \n",
      " 26  EXON                56295 non-null  object \n",
      " 27  INTRON              8803 non-null   object \n",
      " 28  cDNA_position       56304 non-null  object \n",
      " 29  CDS_position        55233 non-null  object \n",
      " 30  Protein_position    55233 non-null  object \n",
      " 31  Amino_acids         55184 non-null  object \n",
      " 32  Codons              55184 non-null  object \n",
      " 33  DISTANCE            108 non-null    float64\n",
      " 34  STRAND              65174 non-null  float64\n",
      " 35  BAM_EDIT            31969 non-null  object \n",
      " 36  SIFT                24836 non-null  object \n",
      " 37  PolyPhen            24796 non-null  object \n",
      " 38  MOTIF_NAME          2 non-null      object \n",
      " 39  MOTIF_POS           2 non-null      float64\n",
      " 40  HIGH_INF_POS        2 non-null      object \n",
      " 41  MOTIF_SCORE_CHANGE  2 non-null      float64\n",
      " 42  LoFtool             60975 non-null  float64\n",
      " 43  CADD_PHRED          64096 non-null  float64\n",
      " 44  CADD_RAW            64096 non-null  float64\n",
      " 45  BLOSUM62            25593 non-null  float64\n",
      "dtypes: float64(12), int64(3), object(31)\n",
      "memory usage: 22.9+ MB\n",
      "CLASS\n",
      "0    48754\n",
      "Name: count, dtype: int64\n",
      "Rows left after deleting CLASS = 1 rows (rows,columns): (48754, 46)\n",
      "Null values for all columns after deleting CLASS = 1: CHROM                   0.000000\n",
      "POS                     0.000000\n",
      "REF                     0.000000\n",
      "ALT                     0.000000\n",
      "AF_ESP                  0.000000\n",
      "AF_EXAC                 0.000000\n",
      "AF_TGP                  0.000000\n",
      "CLNDISDB                0.000000\n",
      "CLNDISDBINCL           99.798991\n",
      "CLNDN                   0.000000\n",
      "CLNDNINCL              99.798991\n",
      "CLNHGVS                 0.000000\n",
      "CLNSIGINCL             99.798991\n",
      "CLNVC                   0.000000\n",
      "CLNVI                  61.555975\n",
      "MC                      1.214259\n",
      "ORIGIN                  0.000000\n",
      "SSR                    99.790786\n",
      "CLASS                   0.000000\n",
      "Allele                  0.000000\n",
      "Consequence             0.000000\n",
      "IMPACT                  0.000000\n",
      "SYMBOL                  0.010256\n",
      "Feature_type            0.010256\n",
      "Feature                 0.010256\n",
      "BIOTYPE                 0.010256\n",
      "EXON                   13.404028\n",
      "INTRON                 86.698527\n",
      "cDNA_position          13.389671\n",
      "CDS_position           14.784428\n",
      "Protein_position       14.784428\n",
      "Amino_acids            14.864421\n",
      "Codons                 14.864421\n",
      "DISTANCE               99.850269\n",
      "STRAND                  0.010256\n",
      "BAM_EDIT               50.709685\n",
      "SIFT                   62.466669\n",
      "PolyPhen               62.497436\n",
      "MOTIF_NAME            100.000000\n",
      "MOTIF_POS             100.000000\n",
      "HIGH_INF_POS          100.000000\n",
      "MOTIF_SCORE_CHANGE    100.000000\n",
      "LoFtool                 6.325635\n",
      "CADD_PHRED              1.973171\n",
      "CADD_RAW                1.973171\n",
      "BLOSUM62               61.332403\n",
      "dtype: float64\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 48754 entries, 0 to 65187\n",
      "Data columns (total 46 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   CHROM               48754 non-null  object \n",
      " 1   POS                 48754 non-null  int64  \n",
      " 2   REF                 48754 non-null  object \n",
      " 3   ALT                 48754 non-null  object \n",
      " 4   AF_ESP              48754 non-null  float64\n",
      " 5   AF_EXAC             48754 non-null  float64\n",
      " 6   AF_TGP              48754 non-null  float64\n",
      " 7   CLNDISDB            48754 non-null  object \n",
      " 8   CLNDISDBINCL        98 non-null     object \n",
      " 9   CLNDN               48754 non-null  object \n",
      " 10  CLNDNINCL           98 non-null     object \n",
      " 11  CLNHGVS             48754 non-null  object \n",
      " 12  CLNSIGINCL          98 non-null     object \n",
      " 13  CLNVC               48754 non-null  object \n",
      " 14  CLNVI               18743 non-null  object \n",
      " 15  MC                  48162 non-null  object \n",
      " 16  ORIGIN              48754 non-null  int64  \n",
      " 17  SSR                 102 non-null    float64\n",
      " 18  CLASS               48754 non-null  int64  \n",
      " 19  Allele              48754 non-null  object \n",
      " 20  Consequence         48754 non-null  object \n",
      " 21  IMPACT              48754 non-null  object \n",
      " 22  SYMBOL              48749 non-null  object \n",
      " 23  Feature_type        48749 non-null  object \n",
      " 24  Feature             48749 non-null  object \n",
      " 25  BIOTYPE             48749 non-null  object \n",
      " 26  EXON                42219 non-null  object \n",
      " 27  INTRON              6485 non-null   object \n",
      " 28  cDNA_position       42226 non-null  object \n",
      " 29  CDS_position        41546 non-null  object \n",
      " 30  Protein_position    41546 non-null  object \n",
      " 31  Amino_acids         41507 non-null  object \n",
      " 32  Codons              41507 non-null  object \n",
      " 33  DISTANCE            73 non-null     float64\n",
      " 34  STRAND              48749 non-null  float64\n",
      " 35  BAM_EDIT            24031 non-null  object \n",
      " 36  SIFT                18299 non-null  object \n",
      " 37  PolyPhen            18284 non-null  object \n",
      " 38  MOTIF_NAME          0 non-null      object \n",
      " 39  MOTIF_POS           0 non-null      float64\n",
      " 40  HIGH_INF_POS        0 non-null      object \n",
      " 41  MOTIF_SCORE_CHANGE  0 non-null      float64\n",
      " 42  LoFtool             45670 non-null  float64\n",
      " 43  CADD_PHRED          47792 non-null  float64\n",
      " 44  CADD_RAW            47792 non-null  float64\n",
      " 45  BLOSUM62            18852 non-null  float64\n",
      "dtypes: float64(12), int64(3), object(31)\n",
      "memory usage: 17.5+ MB\n",
      "Non null counts: None\n",
      "BIOTYPE\n",
      "protein_coding    48738\n",
      "misc_RNA             11\n",
      "Name: count, dtype: int64\n",
      "ORIGIN\n",
      "1      47923\n",
      "33       278\n",
      "3        150\n",
      "5        128\n",
      "17       110\n",
      "9         89\n",
      "25        16\n",
      "49        13\n",
      "0          7\n",
      "32         6\n",
      "41         5\n",
      "35         5\n",
      "4          3\n",
      "513        3\n",
      "16         2\n",
      "65         2\n",
      "37         2\n",
      "69         1\n",
      "129        1\n",
      "12         1\n",
      "13         1\n",
      "29         1\n",
      "27         1\n",
      "43         1\n",
      "85         1\n",
      "19         1\n",
      "11         1\n",
      "57         1\n",
      "2          1\n",
      "Name: count, dtype: int64\n",
      "CLNVC\n",
      "single_nucleotide_variant    45578\n",
      "Deletion                      2057\n",
      "Duplication                    816\n",
      "Indel                          207\n",
      "Insertion                       81\n",
      "Inversion                       13\n",
      "Microsatellite                   2\n",
      "Name: count, dtype: int64\n",
      "Columns left after dropping those in Table 2: Index(['CHROM', 'POS', 'REF', 'ALT', 'AF_ESP', 'AF_EXAC', 'AF_TGP', 'MC',\n",
      "       'IMPACT', 'SYMBOL', 'Feature', 'EXON', 'cDNA_position', 'CDS_position',\n",
      "       'Protein_position', 'Amino_acids', 'Codons', 'STRAND', 'SIFT',\n",
      "       'PolyPhen', 'LoFtool', 'CADD_PHRED', 'BLOSUM62'],\n",
      "      dtype='object')\n",
      "Shape of data frame after dropping those in Table 2 (rows, columns): (48754, 23)\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 39693 entries, 0 to 65187\n",
      "Data columns (total 23 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   CHROM             39693 non-null  object \n",
      " 1   POS               39693 non-null  int64  \n",
      " 2   REF               39693 non-null  object \n",
      " 3   ALT               39693 non-null  object \n",
      " 4   AF_ESP            39693 non-null  float64\n",
      " 5   AF_EXAC           39693 non-null  float64\n",
      " 6   AF_TGP            39693 non-null  float64\n",
      " 7   MC                39681 non-null  object \n",
      " 8   IMPACT            39693 non-null  object \n",
      " 9   SYMBOL            39693 non-null  object \n",
      " 10  Feature           39693 non-null  object \n",
      " 11  EXON              39693 non-null  object \n",
      " 12  cDNA_position     39693 non-null  float64\n",
      " 13  CDS_position      39693 non-null  float64\n",
      " 14  Protein_position  39693 non-null  int64  \n",
      " 15  Amino_acids       39693 non-null  object \n",
      " 16  Codons            39693 non-null  object \n",
      " 17  STRAND            39693 non-null  float64\n",
      " 18  SIFT              18246 non-null  object \n",
      " 19  PolyPhen          18231 non-null  object \n",
      " 20  LoFtool           37220 non-null  float64\n",
      " 21  CADD_PHRED        39454 non-null  float64\n",
      " 22  BLOSUM62          18790 non-null  float64\n",
      "dtypes: float64(9), int64(2), object(12)\n",
      "memory usage: 7.3+ MB\n"
     ]
    }
   ],
   "source": [
    "#### importing packages \n",
    "import numpy as np \n",
    "import pandas as pd \n",
    "\n",
    "#csv to data frame \n",
    "df = pd.read_csv('clinvar_conflicting.csv') \n",
    "df.info()\n",
    "\n",
    "#(1) = objective 1 = removing all conflicting information \n",
    "#deleting all rows where CLASS = 1 \n",
    "df = df[df.CLASS != 1] \n",
    "# make sure everything was deleted \n",
    "unique_values_CLASS = df['CLASS'].value_counts() \n",
    "print(unique_values_CLASS) \n",
    "#check how much the data frame was shrinked \n",
    "print(\"Rows left after deleting CLASS = 1 rows (rows,columns):\", df.shape)\n",
    "#null values percentage \n",
    "null_percentage = df.isnull().mean() * 100 \n",
    "print(\"Null values for all columns after deleting CLASS = 1:\", null_percentage)\n",
    "print(\"Non null counts:\", df.info()) \n",
    "\n",
    "#(2) dropping irrelevant/high empty percentage/extremely low variance variables\n",
    "\n",
    "#getting the number of categories for some columns \n",
    "columns = ['BIOTYPE', 'ORIGIN', 'CLNVC'] \n",
    "for column in columns: \n",
    "    categories = df[column].value_counts() \n",
    "    print(categories)\n",
    "\n",
    "#refer to Table 1 (in the first paper) to see why each column was dropped \n",
    "df = df.drop(['CLASS', 'Consequence', 'CLNDISDB', 'CLNDN', 'CLNVI', 'CLNDISDBINCL', 'CLNDNINCL', 'CLNSIGINCL', 'SSR', 'DISTANCE', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS', 'MOTIF_SCORE_CHANGE', 'INTRON', 'CADD_RAW', 'BAM_EDIT', 'Allele', 'CLNHGVS', 'BIOTYPE', 'Feature_type', 'ORIGIN', 'CLNVC'], axis = 1) \n",
    "print(\"Columns left after dropping those in Table 2:\", df.columns) \n",
    "print(\"Shape of data frame after dropping those in Table 2 (rows, columns):\", df.shape) \n",
    "\n",
    "#(3) = cleaning some errors in data entry in the cDNA position, CDS_position, Protein_position\n",
    "#took some input from chatgpt here, not a direct copy paste \n",
    "problematic = r'\\d+-\\d+|d+-\\?' \n",
    "columns = ['cDNA_position', 'CDS_position', 'Protein_position'] \n",
    "for column in columns: \n",
    "    correct_columns = ~df[column].str.contains(problematic, na=False)\n",
    "    df = df[correct_columns] \n",
    "    df[column] = pd.to_numeric(df[column], errors = 'coerce') \n",
    "    df = df[pd.notna(df[column])]\n",
    "df.info()\n",
    "\n",
    "df.to_csv('clinvar_cleaned.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
